rust/noosphere-ns/src/records.rs

```diff
@@ -0,0 +1,395 @@
+///   // Attenuation must contain a capability with a resource "sphere:{AUD}"
+///   // and action "sphere/publish".
+///   "att": [{
+///     "with": "sphere:did:key:z6MkkVfktAC5rVNRmmTjkKPapT3bAyVkYH8ZVCF1UBNUfazp",
+///     "can": "sphere/publish"
+///   }],
+///   // Additional Ucan proofs needed to validate.
+///   "prf": [],
+///   // Facts contain a single entry with an "address" field containing
+///   // the content address (CID) that the record's sphere's identity maps to.
+///   "fct": [{
+///     "address": "bafy2bzacec4p5h37mjk2n6qi6zukwyzkruebvwdzqpdxzutu4sgoiuhqwne72"
+///   }]
+/// }
+/// ```
+#[derive(Debug, Clone)]
+pub struct NSRecord {
+    /// The wrapped Ucan token describing this record.
+    pub(crate) token: Ucan,
+    /// The resolved content address this record maps to.
+    pub(crate) address: Option<Cid>,
+}
+
+impl NSRecord {
+    /// Creates a new [NSRecord].
+    pub fn new(token: Ucan) -> Self {
+        // Cache the address if "fct" contains an entry matching
+        // the following object without any authority validation:
+        // `{ "address": "{VALID_CID}" }`
+        let mut address = None;
+        for ref fact in token.facts() {
+            if let Value::Object(map) = fact {
+                if let Some(Value::String(addr)) = map.get(&String::from("address")) {
+                    if let Ok(cid) = Cid::from_str(addr) {
+                        address = Some(cid);
+                        break;
+                    }
+                }
+            }
+        }
+
+        Self { token, address }
+    }
+
+    /// Validates the underlying [ucan::Ucan] token, ensuring that
+    /// the sphere's owner authorized the publishing of a new
+    /// content address. Returns an `Err` if validation fails.
+    pub async fn validate<S: Store>(
+        &mut self,
+        store: &SphereDb<S>,
+        did_parser: &mut DidParser,
+    ) -> Result<(), Error> {
+        let identity = self.identity();
+
+        let desired_capability = generate_capability(identity);
+        let proof = ProofChain::from_ucan(self.token.clone(), did_parser, store).await?;
+
+        let mut has_capability = false;
+        for capability_info in proof.reduce_capabilities(&SPHERE_SEMANTICS) {
+            let capability = capability_info.capability;
+            if capability_info.originators.contains(identity)
+                && capability.enables(&desired_capability)
+            {
+                has_capability = true;
+                break;
+            }
+        }
+
+        if !has_capability {
+            return Err(anyhow!("Token is not authorized to publish this sphere."));
+        }
+
+        if self.address.is_none() {
+            return Err(anyhow!(
+                "Missing a valid fact entry with record address. {} {:?}",
+                identity,
+                self.token.facts()
+            ));
+        }
+
+        self.token.check_signature(did_parser).await?;
```

Comment:
This may be useful: https://github.com/ucan-wg/rs-ucan/blob/9ef21415803ff4a64ff4c851adec160dcea2843c/ucan/src/ucan.rs#L92-L95

Reply:
That's used in `NSRecord::is_expired`, used during the "do i need to fetch from network" conditional in `NameSystem::set_record` --  `Ucan::is_expired` could be added to the general `NSRecord::validate` result, but then requires user/sphere interaction to occur at least within the original expiration date, even if several months, otherwise dead links could occur, the rational for *not* including it in the general validation. Although I haven't yet fleshed out the modality of "valid/authenticated" versus "authenticated but expired" and what the desired outcome of a once-active sphere looks like in the long tail, though maybe irrelevant within this context, as there's no gateway to propagate content/records

Reply:
Rolling expiration check into general validation based off of previous discussions, with the NS only propagating valid entries, but still enabling the possibility for e.g. sphere data to make decisions on a record based off of expiration date